tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/experiment_builder.py

```diff
@@ -470,77 +471,106 @@ def _accumulate_all_X_layers(
             assert (col >= 0).all()
-            X_remap = sparse.coo_matrix((X.data, (row, col)), shape=(eb.n_obs, eb.n_var))
+            data = X.data
+            del X
+            gc.collect()
+
             with soma.Experiment.open(eb.experiment_uri, "w", context=SOMA_TileDB_Context()) as experiment:
-                experiment.ms[ms_name].X[layer_name].write(pa.SparseCOOTensor.from_scipy(X_remap))
+                experiment.ms[ms_name].X[layer_name].write(
+                    pa.Table.from_pydict(
+                        {
+                            "soma_dim_0": row,
+                            "soma_dim_1": col,
+                            "soma_data": data,
+                        }
+                    )
+                )
+
+            del row, col, data
             gc.collect()
 
         # Save presence information by dataset_id
         assert dataset.soma_joinid >= 0  # i.e., it was assigned prior to this step
-        pres_data: npt.NDArray[np.bool_] = raw_X.sum(axis=0) > 0
-        if isinstance(pres_data, np.matrix):
-            pres_data = pres_data.A
-        pres_data = pres_data[0]
-        pres_cols: npt.NDArray[np.int64] = local_var_joinids[np.arange(ad.n_vars, dtype=np.int64)]
-
-        assert pres_data.dtype == bool
-        assert pres_cols.dtype == np.int64
-        assert pres_data.shape == (ad.n_vars,)
-        assert pres_data.shape == pres_cols.shape
         assert ad.n_vars <= eb.n_var
-
-        obs_stats, var_stats = _get_axis_stats(raw_X, dataset_obs_joinid_start, local_var_joinids)
+        obs_stats["n_measured_vars"] = (var_stats.nnz > 0).sum()
+        var_stats["n_measured_obs"] = np.zeros(
+            (len(var_stats),), dtype=CENSUS_VAR_TABLE_SPEC.field("n_measured_obs").to_pandas_dtype()
+        )
+        var_stats.n_measured_obs[var_stats.nnz > 0] = ad.n_obs
+
+        # sanity check on stats
+        assert len(var_stats) == ad.n_vars
+        assert len(obs_stats) == ad.n_obs
+        assert (
+            var_stats.n_measured_obs[var_stats.n_measured_obs != 0] == ad.n_obs
+        ).all(), f"n_measured_obs mismatch: {eb.name}:{dataset.dataset_id}"
+        assert (
+            obs_stats.n_measured_vars == np.count_nonzero(var_stats.nnz > 0)
+        ).all(), f"n_measured_vars mismatch: {eb.name}:{dataset.dataset_id}"
 
         results.append(
             (
                 PresenceResult(
                     dataset.dataset_id,
                     dataset.soma_joinid,
                     eb.name,
-                    pres_data,
-                    pres_cols,
+                    (var_stats.nnz > 0).to_numpy(),
+                    var_stats.index.to_numpy(),
                 ),
                 AxisStats(eb.name, obs_stats, var_stats),
             )
         )
 
         # tidy
-        del ad, raw_X, raw_var, local_var_joinids, row, col, X_remap, pres_data, pres_cols, obs_stats, var_stats
+        del ad, local_var_joinids, obs_stats, var_stats
 
     gc.collect()
```

Comment:
indent to gc to free the mem for `ad, local_var_joinids, obs_stats, var_stats` after each iteration?